File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/core/ImageProxy.kt
package dev.hebei.camerax_android.core

import android.graphics.Bitmap
import android.graphics.Rect

class ImageProxy internal constructor(internal val obj: androidx.camera.core.ImageProxy) {
    val format get() = obj.format.imageFormatArgs
    val width get() = obj.width
    val height get() = obj.height
    val planes get() = obj.planes.map { PlaneProxy(it) }.toTypedArray()
    val image
        @androidx.camera.core.ExperimentalGetImage get() = obj.image
    val imageInfo get() = ImageInfo(obj.imageInfo)
    val cropRect get() = obj.cropRect

    fun setCropRect(rect: Rect?) {
        obj.setCropRect(rect)
    }

    fun toBitmap(): Bitmap {
        return obj.toBitmap()
    }

    fun close() {
        obj.close()
    }

    class PlaneProxy(internal val obj: androidx.camera.core.ImageProxy.PlaneProxy) {
        val buffer get() = obj.buffer
        val pixelStride get() = obj.pixelStride
        val rowStride get() = obj.rowStride
    }
}

File: camerax/lib/src/core/image_proxy.dart
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;
import 'package:meta/meta.dart';

final class ImageProxy {
  final $interface.ImageProxy _obj;

  ImageProxy._native(this._obj);

  @internal
  $interface.ImageProxy get obj => _obj;

  $interface.ImageFormat get format => _obj.format;
  int get width => _obj.width;
  int get height => _obj.height;
  $interface.ImageInfo get imageInfo => _obj.imageInfo;
  List<$interface.PlaneProxy> get planes => _obj.planes;

  Future<void> close() => _obj.close();
}

extension ImageProxyObj on $interface.ImageProxy {
  @internal
  ImageProxy get args {
    return ImageProxy._native(this);
  }
}
